feat: int lock and rate limit

xnoname79 opened the pull request:
support bucket getter
add rate limiter
add distrubuted lock
support first time query reset

xnoname79 committed:
feat
support bucket getter
add rate limiter
add distrubuted lock
support first time query reset

TaQuangTu commented on config/config.go:
```diff
@@ -1,15 +1,15 @@
 
 	//postgres
-	POSTGRES_DSN string `env:"POSTGRES_DSN"`
+	POSTGRES_DSN           string `env:"POSTGRES_DSN"`
+	REDIS_LOCKDB_DNS       string `env:"REDIS_LOCKDB_DNS"`
+	REDIS_RATE_LIMITDB_DNS string `env:"REDIS_RATE_LIMITDB_DNS"`
```
DSN not DNS

xnoname79 replied:
Fixed ✅

TaQuangTu commented on migrations/1_create_key_table.sql:
```diff
@@ -1,14 +1,13 @@
     api_key VARCHAR(100) NOT NULL UNIQUE,
     search_engine_id VARCHAR(100) NOT NULL,
-    daily_queries INTEGER DEFAULT 0,
-    daily_limit INTEGER DEFAULT 100,
-    last_used TIMESTAMP WITH TIME ZONE,
-    is_active BOOLEAN DEFAULT TRUE,
-    created_at TIMESTAMP WITH TIME ZONE DEFAULT NOW(),
-    updated_at TIMESTAMP WITH TIME ZONE DEFAULT NOW()
+    daily_queries INTEGER NOT NULL DEFAULT 100,
+    is_active BOOLEAN NOT NULL DEFAULT TRUE,
+    reseted_at TIMESTAMP WITHOUT TIMEZONE NOT NULL DEFAULT NOW(),
+    created_at TIMESTAMP WITHOUT TIMEZONE NOT NULL DEFAULT NOW(),
+    updated_at TIMESTAMP WITHOUT TIMEZONE NOT NULL DEFAULT NOW()
 );
 
 CREATE INDEX idx_api_keys_active ON api_keys(is_active);
```
create another script that modify the table instead.

xnoname79 replied:
Fixed ✅

xnoname79 committed:
update: 1st comment resolve
add new migration file update for key_table
fix typo DNS -> DSN
remove count(*) records for paritions: not neccesary
fix query update daily reset

TaQuangTu committed:
feat: add string utility

TaQuangTu committed:
fix: complete redis basic usage

TaQuangTu committed:
docs: add guide generating grpc messages

TaQuangTu committed:
ci: setup precommit

TaQuangTu committed:
ci: fix precommit install and add to docs

TaQuangTu committed:
style: lint (no verify)

TaQuangTu commented on internal/server/server.go:
```diff
@@ -15,12 +17,17 @@ type SearchServer struct {
 }
 
-func NewSearchServer(dsn string) *SearchServer {
+func NewSearchServer(conf *config.Config, lockDb lockdb.ILockDb, rateLimiter lockdb.RateLimiter) *SearchServer {
```
Why not using pointer for these parameters? AFAIK, it could help avoid unpredicted behaviors of shallow copy.

xnoname79 replied:
They are defined as interfaces, so they were passed in as pointer from the main thread.

TaQuangTu commented on internal/search/key_manager.go:
```diff
@@ -3,21 +3,28 @@ package search
 }
 
-func NewApiKeyManager(dsn string) *ApiKeyManager {
+func NewApiKeyManager(dsn string, lockDb lockdb.ILockDb, rateLimiter lockdb.RateLimiter) *ApiKeyManager {
```
Better to use pointer params?

xnoname79 replied:
They are defined as interfaces, so they were passed in as pointer from the main thread.

TaQuangTu commented on internal/server/server.go:
```diff
@@ -29,18 +36,51 @@ func (s *SearchServer) Search(ctx context.Context, req *protogen.SearchRequest)
+				return
+			}
+			defer ourLock.Unlock()
+			s.keyManager.ResetDailyCounts(100, availableKey.ResetedAt)
```
If this reset statement failed, the key will be no longer eligible to use. We may need a better way for reseting the daily quota.

TaQuangTu committed:
ci: fix go lint, go import, go vet

TaQuangTu committed:
ci: add quality check workflow

TaQuangTu committed:
ci: fix default branch name

TaQuangTu committed:
Merge branch 'master' into feat/support-bucket-getter
